Userland/Libraries/LibWeb/Page/EventHandler.cpp

```diff
@@ -333,8 +333,7 @@ EventResult EventHandler::handle_mouseup(CSSPixelPoint viewport_position, CSSPix
                     JS::NonnullGCPtr<DOM::Document> document = *m_navigable->active_document();
                     auto href = link->href();
                     auto url = document->parse_url(href);
-
-                    if (button == UIEvents::MouseButton::Primary && (modifiers & UIEvents::Mod_PlatformCtrl) != 0) {
+                    if (button == UIEvents::MouseButton::Primary) {
                         m_navigable->page().client().page_did_click_link(url, link->target().to_byte_string(), modifiers);
```

Comment:
What happens now when we click a normal link? IIRC non-ctrl clicks go through the anchor element's activation behavior, so does this cause that navigation to happen twice?

Reply:
It does navigate twice, though I don't think for the reason you mention here. The problem is all the chromes still call `load_url` within there handler for `on_link_click`, which has been (partially) dead code for a while now. It's made slightly awkward by some questionable reuse of the `on_link_click` callback too (it's called by places other than the web content client, which means in some places may still needs the `load_url` behaviour.

Reply:
I've pushed a fix for Ladybird and Browser (can't really do the macOS stuff, since I don't have mac).

Reply:
If I were to do the mac bits, what would I do to test I haven't broken anything (or rather, that I fixed something)?

Reply:
I think the changes needed are just:

The `on_link_click` callback should only do anything when the control modifier is set (to match that previously LibWeb only sent this event when control is held).
https://github.com/SerenityOS/serenity/blob/970c02574135598b609ed9e29c8796f5c06f84b7/Ladybird/AppKit/UI/LadybirdWebView.mm#L616-L628

This `openLink`  function should direly call `loadURL`  rather than reuse the `on_link_click` callback.
https://github.com/SerenityOS/serenity/blob/970c02574135598b609ed9e29c8796f5c06f84b7/Ladybird/AppKit/UI/LadybirdWebView.mm#L1221-L1224

---

Then to check if "openLink" works, try a context menu event (like "Open Image"). For the `on_link_click` callback, just try clicking any link (and check it opens without requesting the page twice).

Reply:
Back again after a while, I've made a new PR #25857, which is a little more general, and does not require updating the browser chromes (so avoids any messing with macOS stuff).